### PR TITLE
refactor(timeline creation): idempotency checking

### PR DESCRIPTION
**Pull request by problame**
# Context

In the PGDATA import code (https://github.com/neondatabase/neon/pull/9218) I add a third way to create timelines, namely, by importing from a copy of a vanilla PGDATA directory in object storage.

For idempotency, I'm using the PGDATA object storage location specification, which is stored in the IndexPart for the entire lifespan of the timeline. When loading the timeline from remote storage, that value gets stored inside `struct Timeline` and timeline creation compares the creation argument with that value to determine idempotency of the request.

# Changes

This PR refactors the existing idempotency handling of Timeline bootstrap and branching such that we simply compare the `CreateTimelineIdempotency` struct, using the derive-generated `PartialEq` implementation.

Also, by spelling idempotency out in the type names, I find it adds a lot of clarity.

The pathway to idempotency via requester-provided idempotency key also becomes very straight-forward, if we ever want to do this in the future.

# Refs
* platform context: https://github.com/neondatabase/neon/pull/9218
* product context: https://github.com/neondatabase/cloud/issues/17507
* stacks on top of https://github.com/neondatabase/neon/pull/9366

### CODE REVIEW

**Review by VladLazar — APPROVED** [review]
Looks good on top of https://github.com/neondatabase/neon/pull/9366

File: pageserver/src/tenant.rs [hunk 1]
```diff
@@ -3951,28 +4036,96 @@ impl Tenant {
...
+                    .remote_client
+                    .wait_completion()
+                    .await
+                    .context("wait for timeline uploads to complete")?;
+
+                // TODO: shouldn't we also wait for timeline to become active?
+                // Code before this(https://github.com/neondatabase/neon/pull/9366) refactoring
+                // didn't do it.
+
+                Ok(StartCreatingTimelineResult::Idempotent(existing))
```
Comment (problame) [inline pageserver/src/tenant.rs:4126]: Hm, about this piece of code here:

1. Can we rely on the callers to do wait_completion()? The not-cfg(test) code always waits for completion already.
2. About the `TODO :shouldn't we also wait...`: this would probably also make more sense in the place that does the activation in the case of `StartCreatingTimelineResult::Created`.

Comment (VladLazar) [reply pageserver/src/tenant.rs]: > Hm, about this piece of code here:
> 
>     1. Can we rely on the callers to do wait_completion()? The not-cfg(test) code always waits for completion already.

Yes, that would be my expectation.

>     2. About the `TODO :shouldn't we also wait...`: this would probably also make more sense in the place that does the activation in the case of `StartCreatingTimelineResult::Created`.

Yeah, I think it makes sense to move to the place where we do the activation.

Question though:
For `StartCreatingTimelineResult::Idempotent`, why do we call `wait_completion` on the remote client twice?
Once in `start_creating_timeline` and once in `create_timeline`?

File: pageserver/src/tenant.rs [hunk 2]
```diff
@@ -4032,62 +4082,30 @@ impl Tenant {
...
-                        if existing.get_ancestor_timeline_id().is_some() {
-                            info!("timeline already exists with an ancestor");
-                            return Err(CreateTimelineError::Conflict);
-                        }
-                        if existing.get_ancestor_lsn() != Lsn::INVALID {
-                            info!("timeline already exists with an ancestor LSN");
+            Err(TimelineExclusionError::AlreadyExists { existing, arg }) => {
+                {
+                    let existing = &existing.create_idempotency;
+                    let _span = info_span!("idempotency_check", ?existing, ?arg).entered();
```
Comment (arpad-m) [inline pageserver/src/tenant.rs:4088]: the debug impl of a timeline just prints the timeline ID, can't we attach the timline ID to the span instead?

File: pageserver/src/tenant.rs [hunk 3]
```diff
@@ -4020,7 +4070,7 @@ impl Tenant {
         new_timeline_id: TimelineId,
         idempotency: CreateTimelineIdempotency,
     ) -> Result<StartCreatingTimelineResult<'_>, CreateTimelineError> {
-        match self.create_timeline_create_guard(new_timeline_id) {
+        match self.create_timeline_create_guard(new_timeline_id, idempotency) {
```
Comment (arpad-m) [inline pageserver/src/tenant.rs:4073]: Why did you add the param here? It's passed on by the function. we could instead just use `idempotency` directly instead of `arg` in the match case below.
